### PR TITLE
[BugFix] Fix bugs when materialized view with sort keys

### PR DESCRIPTION
**Pull request by LiShuMing**
When the materialized view has set sort keys,  or adjust output column orders by the default sort key policy, need to set target names by the query's output columns' order in refreshing materialized view.

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 30 / 32 (93.75%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|11|13|84.62%|[446, 447]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateMaterializedViewStatement.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/InsertStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 1]
```diff
@@ -386,6 +386,17 @@ public void setLastRefreshTime(long lastRefreshTime) {
...
 
+
+    // Materialized view's output columns may be different from defined query's output columns.
+    // Record the indexes based on materialized view's column output.
+    // eg: create materialized view mv as select col1, col2, col3 from tbl
+    //  desc mv             :  col2, col1, col3
+    //  queryOutputIndexes  :  1, 0, 2
+    // which means 0th of query output column is in 1th mv's output columns, and 1th -> 0th, 2th -> 2th.
+    @SerializedName(value = "queryOutputIndexes")
+    protected List<Integer> queryOutputIndexes = Lists.newArrayList();
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:397]: https://github.com/StarRocks/starrocks/pull/31702. dose this field cause the similar problem above?

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java [hunk 2]
```diff
@@ -261,6 +261,17 @@ public Void visitCreateMaterializedViewStatement(CreateMaterializedViewStatement
...
             statement.setMvColumnItems(mvColumns);
+            Map<Integer, Integer> queryToMVOutputIndexMap = Maps.newHashMap();
+            for (int i = 0; i < mvColumnPairs.size(); i++) {
+                Pair<Column, Integer> pair = mvColumnPairs.get(i);
+                queryToMVOutputIndexMap.put(pair.second, i);
+            }
+            List<Integer> queryOutputIndexes = Lists.newArrayList();
+            for (int i = 0; i < mvColumnPairs.size(); i++) {
+                Preconditions.checkState(queryToMVOutputIndexMap.containsKey(i));
+                queryOutputIndexes.add(queryToMVOutputIndexMap.get(i));
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java:272]: put into a map<int, int> and get from the map, 

is it equivalent to sort the zipWithIndex().sorted(Comparator::comparaing(Pair::getFirst)) ?

**Review by murphyatwork — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java (hunk 1 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:397]: no. It should not.

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

coverage check fail. please retry. :fearful:

[Please let me know](https://github.com/naver/cover-checker/issues/new) when error again.

java.io.IOException: Server returned HTTP response code: 500 for URL: https://api.github.com/repos/StarRocks/starrocks/issues/33125/comments

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 30 / 32 (93.75%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|11|13|84.62%|[446, 447]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateMaterializedViewStatement.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/InsertStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|1|100.00%|[]|

**Review by Astralidea — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 26 / 28 (92.86%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|7|9|77.78%|[449, 450]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateMaterializedViewStatement.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/InsertStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by ABingHuang — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**